connector/connect/common/src/main/protobuf/spark/connect/base.proto

```diff
@@ -183,6 +183,60 @@ message ExecutePlanResponse {
   }
 }
 
+// Request to transfer client-local artifacts.
+message AddArtifactsRequest {
+
+  // Definition of an Artifact.
+  message Artifact {
+    // The name of the artifact is expected in the form of a "Relative Path" that is made up of a
+    // sequence of directories and the final file element.
+    // Examples of "Relative Path"s: "jars/test.jar", "classes/xyz.class", "abc.xyz", "a/b/X.jar".
+    // The server is expected to maintain the hierarchy of files as defined by their name. (i.e
+    // The relative path of the file on the server's filesystem will be the same as the name of
+    // the provided artifact)
+    string name = 1;
+    // Raw data.
+    bytes data = 2;
+  }
```

Comment:
Should consider allowing clients to specify artifacts from Ivy coordinates, either in addition or instead of directly providing the contents of a JAR? This seems to be a cleaner separation of client and server; the client shouldn't need to have a copy of the JARs it will use to execute some UDF. Spark already has support for localizing artifacts from an external repository.

Reply:
We can add it. Ivy coordinates become a bit of a hassle when someone is using a private repository. Quite a few of our customers provides jars as is, they don't want to or can't manage repositories for various reasons. Furthermore we also need to deal with REPL generated code, and there are a bunch of other use cases that still need this kind of synchronization.

Reply:
Good point; you have definitely convinced me that supporting the raw data xfer as specified here is needed. I do think we should consider Ivy coordinate support as a future extension, at least, and make sure we can support it here. Right now `bytes data` is a required field, so we may need to adjust that to support future ways of specifying the required content?

Reply:
The introduction of `SingleChunkArtifact`, `BeginChunkedArtifact` etc helps with extensibility because we can now add a new message type, say "IvyArtifact", and include it in the `payload` of `AddArtifactsRequest`.